### ISSUE TITLE
Changing the root tree live

### ISSUE THREAD

**Issue by Yomguithereal** [Yomguithereal/baobab-react#67]

**Comment by georgehenderson** [comment]
Are there any workarounds for this issue? Is it possible to manually cleanup a tree and then reset with a new root cursor. Currently running into memory issues with SSR, and recreating the tree for each request is not really an option without a larger refactor. Problem is very similar to the discussion that started here https://github.com/Yomguithereal/baobab/pull/135

**Comment by Yomguithereal** [comment]
What do you mean by cleanup the tree? Isn't `tree.set()` not enough to wipe the associated data?

**Comment by georgehenderson** [comment]
There seems to be a memory leak when just resetting the tree using `tree.set()`. Should that remove all of the old event listeners too, like release does?

**Comment by Yomguithereal** [comment]
`tree.set()` won't remove the old event listeners only `release` does. But `release` is a bit terminal as you won't be able to use the same tree anymore after. Why is creating a tree with each request costly? Do you hit performance issues when creating the tree?

**Comment by georgehenderson** [comment]
Its not costly from a performance standpoint, more from a application refactor perspective. Currently I am exporting my tree in a module that the rest of my application is referencing directly. Similar to 
```
module.exports = new Baobab()
```
Ideally there would be some way for me to `tree.release()` then `tree.set(initialState)` so I can reuse the ref.

**Comment by Yomguithereal** [comment]
Are you in fact searching to cleanup the listeners & resetting the data without killing the tree?

**Comment by Yomguithereal** [comment]
It's a bit of a hack but you can try the following (taken from the `kill` method itself but skipping the nasty parts you don't want):

```js
for (k in tree._cursors)
  tree._cursors[k].release();
delete tree._cursors;
```

**Comment by georgehenderson** [comment]
yes, something similar to what christianalfoni recommended with `tree.mute() // removes listeners`.

**Comment by Yomguithereal** [comment]
The equivalent to `tree.mute()` would be `tree.unbindAll()` or `tree.off('update')`. but you should probably use the solution I posted above before.

**Comment by georgehenderson** [comment]
Great! thanks will give it a try.

**Comment by Yomguithereal** [comment]
Now that I think of it, there might be an issue with the react integration since I fear the `componentWillUnmount` hooks are not triggered server-side.

**Comment by georgehenderson** [comment]
Yeah, I'm seeing issues with this method. It looks like its from old selections.

**Comment by georgehenderson** [comment]
What are the down sides to using `tree.unbindAll()`? Will that still leave around old data if we then call `tree.set()`. Seems to work but can't tell if there is still a memory issue, yet.

**Comment by Yomguithereal** [comment]
There shouldn't be a downside to `unbindAll` (maybe edge cases if using recursive monkeys).

**Comment by Yomguithereal** [comment]
If you stumble upon a memory issue, just tell me and we'll investigate it together.

**Comment by georgehenderson** [comment]
Great thank you for your help and contributions to OS, baobab is bad ass.
